plugins/inputs/jenkins/jenkins.go

```diff
@@ -0,0 +1,440 @@
+		return
+	}
+	var wg sync.WaitGroup
+	for _, job := range jobs {
+		wg.Add(1)
+		go func(name string, wg *sync.WaitGroup, acc telegraf.Accumulator) {
+			defer wg.Done()
+			if te := j.getJobDetail(jobRequest{
+				name:    name,
+				parents: []string{},
+				layer:   0,
+			}, acc); te != nil {
+				acc.AddError(te)
+			}
+		}(job.Name, &wg, acc)
+	}
+	wg.Wait()
+}
+
+// wrap the tcp request with doGet
+// block tcp request if buffered channel is full
+func (j *Jenkins) doGet(tcp func() error) error {
+	j.semaphore <- struct{}{}
+	if err := tcp(); err != nil {
+		<-j.semaphore
+		return err
+	}
+	<-j.semaphore
+	return nil
+}
+
+func (j *Jenkins) getJobDetail(sj jobRequest, acc telegraf.Accumulator) error {
+	if j.MaxSubJobDepth > 0 && sj.layer == j.MaxSubJobDepth {
+		return nil
+	}
+	// filter out excluded job.
+	if j.jobFilter != nil && j.jobFilter.Match(sj.hierarchyName()) {
+		return nil
+	}
+	url := j.URL + "/job/" + strings.Join(sj.combined(), "/job/") + "/api/json"
+	var jobDetail *gojenkins.Job
+	var err error
+	err = j.doGet(func() error {
+		jobDetail, err = j.instance.GetJob(sj.name, sj.parents...)
+		return err
+	})
+	if err != nil {
+		return fmt.Errorf("error retrieving inner jobs[%s]: ", url)
+	}
+
+	var wg sync.WaitGroup
+	for k, innerJob := range jobDetail.Raw.Jobs {
+		if k < len(jobDetail.Raw.Jobs)-j.MaxSubJobPerLayer-1 {
+			continue
+		}
+		wg.Add(1)
+		// schedule tcp fetch for inner jobs
+		go func(innerJob gojenkins.InnerJob, sj jobRequest, acc telegraf.Accumulator) {
+			defer wg.Done()
+			if te := j.getJobDetail(jobRequest{
+				name:    innerJob.Name,
+				parents: sj.combined(),
+				layer:   sj.layer + 1,
+			}, acc); te != nil {
+				acc.AddError(te)
+			}
+		}(innerJob, sj, acc)
+	}
+	wg.Wait()
+
+	// collect build info
+	number := jobDetail.Raw.LastBuild.Number
+	if number < 1 {
+		// no build info
+		return nil
+	}
+	baseURL := "/job/" + strings.Join(sj.combined(), "/job/") + "/" + strconv.Itoa(int(number))
+	// jobDetail.GetBuild is not working, doing poll directly
+	build := &gojenkins.Build{
+		Jenkins: j.instance,
+		Depth:   1,
+		Base:    baseURL,
+		Raw:     new(gojenkins.BuildResponse),
+	}
+	var status int
+	err = j.doGet(func() error {
+		status, err = build.Poll()
+		return err
+	})
+	if err != nil || status != 200 {
+		if err == nil && status != 200 {
+			err = fmt.Errorf("status code %d", status)
+		}
+		return fmt.Errorf("error retrieving inner jobs[%s]: %v", j.URL+baseURL+"/api/json", err)
+	}
+
+	if build.Raw.Building {
+		log.Printf("D! Ignore running build on %s, build %v", sj.name, number)
+		return nil
+	}
+
+	// stop if build is too old
+
+	if (j.MaxBuildAge != internal.Duration{Duration: 0}) {
+		buildAgo := time.Now().Sub(build.GetTimestamp())
+		if buildAgo.Seconds() > j.MaxBuildAge.Duration.Seconds() {
+			log.Printf("D! Job %s build %v too old (%s ago), skipping to next job", sj.name, number, buildAgo)
+			return nil
+		}
+	}
+
+	gatherJobBuild(sj, build, acc)
+	return nil
```

Comment:
Does this only get the latest build for each job?

Reply:
Yes